File: 300-LongestIncreasingSubsequence/300-LongestIncreasingSubsequence.java
/*

300. Longest Increasing Subsequence
Medium

Given an unsorted array of integers, find the length of longest increasing subsequence.

Example:

Input: [10,9,2,5,3,7,101,18]
Output: 4 
Explanation: The longest increasing subsequence is [2,3,7,101], therefore the length is 4. 

Note:

    There may be more than one LIS combination, it is only necessary for you to return the length.
    Your algorithm should run in O(n2) complexity.

Follow up: Could you improve it to O(n log n) time complexity?

*/

class Solution {
    public int lengthOfLIS(int[] nums) {
        Integer[][] cache = new Integer[nums.length + 1][nums.length];
        return dfs(cache, nums, -1, 0);
    }
    private int dfs(Integer[][] cache, int[] nums, int prev, int idx){
        if(idx >= nums.length) return 0;
        if(cache[prev + 1][idx] != null) return cache[prev + 1][idx];
        int take = 0;
        if(prev < 0 || nums[idx] > nums[prev])
            take = 1 + dfs(cache, nums, idx, idx + 1);
        int skip = dfs(cache, nums, prev, idx + 1);
        return cache[prev + 1][idx] = Math.max(take, skip);
        
    }
}

File: 300-LongestIncreasingSubsequence/300-LongestIncreasingSubsequence.py
"""

300. Longest Increasing Subsequence

Given an integer array nums, return the length of the longest strictly increasing

Example 1:

Input: nums = [10,9,2,5,3,7,101,18]
Output: 4
Explanation: The longest increasing subsequence is [2,3,7,101], therefore the length is 4.

Example 2:

Input: nums = [0,1,0,3,2,3]
Output: 4

Example 3:

Input: nums = [7,7,7,7,7,7,7]
Output: 1

 

Constraints:

    1 <= nums.length <= 2500
    -104 <= nums[i] <= 104

Follow up: Can you come up with an algorithm that runs in O(n log(n)) time complexity?

"""

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        def dfs(idx, prev):
            if idx >= len(nums): return 0
            if (idx, prev) in cache: return cache[(idx, prev)]
            take, skip = 0, 0
            if prev == -1 or nums[idx] > nums[prev]:
                take = 1 + dfs(idx + 1, idx)
            skip = dfs(idx + 1, prev)
            cache[(idx, prev)] = max(take, skip)
            return cache[(idx, prev)]
        cache = {}
        return dfs(0, -1)
